### DISCUSSION TITLE
Ability matchers

### DISCUSSION THREAD

**Discussion by mrazauskas** [tstyche/tstyche#224] [General]
In some cases user must use `.toRaiseError()` matcher to test something. This looks wrong. These tests are not written from user perspective, also indirect tests are unclear for reader.

This is because TSTyche is missing the ability matchers:

- [x] `.toHaveProperty()`, #36
- [x] `.toBeCallableWith()`, #447
- [x] `.toBeConstructableWith()`, #451
- [x] `.toBeInstantiableWith()` (checks if a generic type, interface, function or class can be instantiated with given type arguments), #399
- [x] `.toBeApplicable` (to test decorator applicability), #100
- [x] `.toAcceptProps()` (checks if a component accepts given props), #237

---

Did you find another cases there `.toRaiseError()` matcher is still needed? Leave a message, please.

**Comment by meduzen** [comment]
Hi @mrazauskas ! Seeing [your PR](https://github.com/frontacles/frontacles/pull/57) in my slowly-progressing library, I have a question regarding `toBeConstructableWith`.

_(I prefer to post it here instead of in the PR since it will be more searchable for other persons.)_

Is there any difference or nuance between what this matcher achieves, and an _expect error_ matcher? Are there use case where `expect(SomeClass).type.not.toBeConstructableWith(invalidParams)` would be true, but not throw any error?

Thinking about these examples:

```js
class ClassThatExplicitlyThrows {
  constructor(params) {
    if (invalid(params)) {
      throw new TypeError('ClassThatExplicitlyThrows constructor: invalid params.')
    }

    // do normal things…
  }
}

class ClassThatSilentlyFails {
  constructor(params) {
    if (invalid(params)) {
      return
    }

    // do normal things…
  }
}

class ClassWithoutValidation {
  constructor(params) {
    // try to do normal things, and crashes if invalid params
  }
}

```

Does it change anything It the class implementation has or doesn't have such validation mechanism, or if it fails by itself, or if it doesn’t do anything at all?

**Reply by mrazauskas** [reply]
Thanks. This is very good question. Might be I should add more details to documentation.

Both `.toBeConstructableWith()` and `.toBeCallableWith()` first of all make sure that the expression is actually constructable or callable.

One more detail, even before that the file is type-checked and any errors get reported.

The last step is to erase the matcher syntax and to type-check the file again. If there are new errors the expression is reported as not constructable or callable. The syntax is erased roughly like this:

```ts
expect(getPerson).type.not.toBeCallableWith({ name: false })

;     (getPerson)                          ({ name: false })         
```

---

The main difference from expecting an error is that the error message is treated as implementation detail. The shape of messages is changing between TypeScript version. The ability marchers make it easier to test against several TS versions.

---

The of the class is interesting aspect. TSTyche does not execute the code. So none of the matchers can see `throw` statements.

Only if you use matchers in unit tests that are executed you will see difference. The `.not.toBeConstructable()` allows executing the code without invoking the class constructor. In contrary, using error matcher a function or a constructor must be invoked.

These are the differences that came to my mind quickly.
